splink/comparison_level_composition.py

```diff
@@ -0,0 +1,262 @@
+            Defaults to None.
+
+    Examples:
+        >>> import splink.duckdb.duckdb_comparison_level_library as cll
+        >>> # *Not* a null on first name `first_name`
+        >>> cll.cl_not(cll.exact_match("first_name"))
+
+        >>> import splink.duckdb.duckdb_comparison_level_library as cll
+        >>> # Find all exact matches *not* on the first of January
+        >>> dob_first_jan =  {
+        >>>    "sql_condition": "SUBSTR(dob_std_l, -5) = '01-01'",
+        >>>    "label_for_charts": "Date is 1st Jan",
+        >>> }
+        >>> exact_match_not_first_jan = cll.cl_and(
+        >>>     cll.exact_match_level("dob"),
+        >>>     cll.cl_not(dob_first_jan),
+        >>>     label_for_charts = "Exact match and not the 1st Jan"
+        >>> )
+
+
+    Returns:
+        ComparisonLevel
+            A new ComparisonLevel with the negated SQL condition and label_for_charts
+    """
+    dicts, sql_dialect = _parse_comparison_levels(cll)
+    result = {}
+    cld = dicts[0]
+    result["sql_condition"] = f"NOT ({cld['sql_condition']})"
+
+    result["label_for_charts"] = (
+        label_for_charts if label_for_charts else f"NOT ({_label_for_charts(cld)})"
+    )
+
+    if m_probability:
+        result["m_probability"] = m_probability
+
+    return ComparisonLevel(result, sql_dialect=sql_dialect)
+
+
+def _cl_merge(
+    *clls: ComparisonLevel | dict,
+    clause: str,
+    label_for_charts=None,
+    m_probability=None,
+    is_null_level=None,
+) -> ComparisonLevel:
+    if len(clls) == 0:
+        raise ValueError("Must provide at least one ComparisonLevel")
+
+    dicts, sql_dialect = _parse_comparison_levels(*clls)
+    result = {}
+    conditions = ("(" + d["sql_condition"] + ")" for d in dicts)
+    result["sql_condition"] = f" {clause} ".join(conditions)
+
+    # Set to null level if all supplied levels are "null levels"
+    if is_null_level is None:
+        if all([d.setdefault("is_null_level", False) for d in dicts]):
+            result["is_null_level"] = True
+
+    if label_for_charts:
+        result["label_for_charts"] = label_for_charts
+    else:
+        labels = ("(" + _label_for_charts(d) + ")" for d in dicts)
+        result["label_for_charts"] = f" {clause} ".join(labels)
+
+    if m_probability:
+        result["m_probability"] = m_probability
+
+    return ComparisonLevel(result, sql_dialect=sql_dialect)
+
+
+def _label_for_charts(comparison_dict: dict) -> str:
+    backup = comparison_dict["sql_condition"]
+    label = comparison_dict.get("label_for_charts", backup)
+
+    colname = comparison_dict.get("column_name")
+    if colname is None:
+        return label
+    # if null level, prefix with the column name
+    if comparison_dict.get("is_null_level", False):
+        label = f"{colname} is {label.upper()}"
+    # if exact match, suffix w/ colname
+    elif label.lower() == "exact match":
+        label = f"{label} on {colname}"
+
+    return label
+
+
+def _parse_comparison_levels(
```

Comment:
Please feel free to write something up on this if you get the chance this evening. If not, I'll do some work on this tomorrow.

I don't imagine it will be overly difficult. `as_dict()` (which you're using to extract the dictionaries) is simply a method on `ComparisonLevel`. You'd simply need to cross reference to that and make some adjustments to how we're evaluating the CLs.

Reply:
I can come through and make these tweaks if you want, but I don't want to be editing at the same time as you or we'll get merge conflicts. If you let me know when you're done tweaking then I can take a pass, so we never are working on it at the same time :)

Reply:
Of course. I'll upload push some more code in the next hour or so for this and then you're free to make whatever adjustments you want from there.